### PR TITLE
Allow email access list overrides for web buckets

### PR DESCRIPTION
**Pull request by lgruen**
If someone isn't in the generic access list for the web bucket, they can be listed in a `.access` file in the first subdirectory instead. This is helpful to give external collaborators access to particular files.

Context: https://centrepopgen.slack.com/archives/C018KFBCR1C/p1648103614839679?thread_ts=1647580532.711369&cid=C018KFBCR1C

### CODE REVIEW

File: web/main.py [hunk 1]
```diff
@@ -43,17 +62,32 @@ def handler(dataset=None, filename=None):
...
     members = read_secret(
         dataset_project_id, f'{dataset}-web-access-members-cache'
     ).split(',')
+
     if email not in members:
-        logger.warning(f'{email} is not a member of the {dataset} access group')
-        abort(403)
+        # Second chance: if there's a '.access' file in the first subdirectory,
+        # check if the email is listed there.
+        split_subdir = filename.split('/', maxsplit=1)
```
Comment (illusional) [inline web/main.py:76]: I think using os.path.dirname is cleaner for the downstream code
```suggestion
        subdir = os.path.dirname(filename)
```

**Review by MattWellie — APPROVED** [review]

File: web/main.py [hunk 2]
```diff
@@ -43,17 +62,32 @@ def handler(dataset=None, filename=None):
...
+        if len(split_subdir) == 2 and split_subdir[0]:
+            access_list_filename = f'{split_subdir[0]}/.access'
```
Comment (illusional) [inline web/main.py:78]:
```suggestion
            access_list_filename = os.path.join(subdir, '.access')
```

File: web/main.py [hunk 3]
```diff
@@ -43,17 +62,32 @@ def handler(dataset=None, filename=None):
...
+            blob = bucket.get_blob(access_list_filename)
+            if blob is None:
+                logger.warning(f'{email} is not a member of the {dataset} access group')
+                abort(403)
+            access_list = blob.download_as_string().decode('utf-8').splitlines()
```
Comment (illusional) [inline web/main.py:83]: Does `download_as_string` download it as binary, you've got a decode here which seems strange

**Review by illusional — APPROVED** [review]
Kind of awkward to implement, but I think a `.access` file only giving you access to the current dir is going to get confusing.

File: web/main.py (hunk 3 above)
Comment (lgruen) [reply web/main.py:83]: Yes, indeed. This has been fixed with newer versions of the library, but unfortunately `hail` (which `cpg-utils` depends on) pins an old version of `google-cloud-storage`...

File: web/main.py (hunk 1 above)
Comment (lgruen) [reply web/main.py:76]: I only want the first subdirectory down from the top-level though, not the `.access` file further down. I thought about allowing those files at any subdirectory level, but then you'd have to walk up the tree and check for blob existence on every level.

File: web/main.py (hunk 2 above)
Comment (lgruen) [reply web/main.py:78]: At the moment, `split_subdir[0]` can't contain any slashes, so `os.path.join` doesn't really help. Would be cleaner if we used `dirname` as you suggested, but that would have different semantics.

File: web/main.py (hunk 3 above)
Comment (MattWellie) [reply web/main.py:83]: This was a fun one to find out 🤦 the d_as_string downloads bytes, and the d_as_bytes doesn't exist 🎉

**Pull request merged**